---
title: "coreExome selection analysis"
author: ''
date: '`r format(Sys.Date())`'
output:
  html_document:
  toc: yes
--- 
  
```{r setup, echo = FALSE, message=FALSE}
library(data.table)
library(dplyr)
library(ggplot2)
library(GenomicRanges)
library(glida)
library(VariantAnnotation)
library(TxDb.Hsapiens.UCSC.hg19.knownGene) # for annotation
library(org.Hs.eg.db)  # to convert from Entrez Gene ID to Symbol
data_dir <- '/media/xsan/scratch/merrimanlab/murray/working_dir/coreExome_selection/NZ/'


get_faw <- function(faw_path, pop){
  faw <- list()
  for(i in 1:22){
    a= read.table(paste0(faw_path,pop,i,'.faw'), header=FALSE, skip=5)
    names(a)=c("RefStart","Refend","RefMid","chrom_start","chrom_end","Midpoint","NumSites","Missing","S","Eta","Eta_E","Pi","FuLi_D","FuLi_F","FayWu_H")
    b=a[a$FayWu_H != "NaN",]
    b$POP=as.factor(pop)
    b$chrom <- paste0('chr',i)
    faw[[i]] <- b[,c("chrom","chrom_start","chrom_end", "FayWu_H","POP")]
  }
  return(do.call(rbind,faw))
  
}


get_td <- function(td_path, pop){
  td <- list()
  for(i in 1:22){
    a= read.table(paste0(td_path,pop,i,'.taj_d'), header=TRUE)
    names(a)=c("chrom", "chrom_start","NumSites","TajimaD")
    b=a[a$TajimaD != "NaN",]
    b$POP=as.factor(pop)
    b$chrom <- paste0('chr',i)
    b$chrom_end <- b$chrom_start + b[2,"chrom_start"] - b[1,"chrom_start"]
    td[[i]] <- b[,c("chrom","chrom_start","chrom_end", "TajimaD","POP")]
  }
  return(do.call(rbind,td))
  
}

get_nsl <- function(nsl_path, pop){
  nsl <- list()
  for(i in 1:22){
    a= read.table(paste0(nsl_path,pop,'_',i,'.nsl.out.100bins.norm'), header=FALSE)
    names(a)=c("marker","chrom_start","freq_1","sL1","sL0","unstd_nsl", "std_nsl", "significant")
    a$POP=as.factor(pop)
    a$chrom <- paste0('chr',i)
    a$chrom_end <- a$chrom_start + 1
    nsl[[i]] <-a[,c("chrom","chrom_start","chrom_end", "std_nsl","POP", "significant")]
  }
  return(do.call(rbind,nsl))
  
}


get_ihs <- function(ihs_path, pop){
  ihs <- list()
  for(i in 1:22){
    a= read.table(paste0(ihs_path,pop,'_',i,'.ihs.out.100bins.norm'), header=FALSE)
    names(a)=c("marker","chrom_start","freq_1","ihh1","ihh0","unstd_ihs", "std_ihs", "significant")
    a$POP=as.factor(pop)
    a$chrom <- paste0('chr',i)
    a$chrom_end <- a$chrom_start + 1
    ihs[[i]] <- a[,c("chrom","chrom_start","chrom_end", "std_ihs","POP", "significant")]
  }
  return(do.call(rbind,ihs))
  
}


get_kaks <- function(kaks_path, pop){
  kaks <- list()
  for(i in 1:22){
    a= read.table(paste0(kaks_path,pop,i,'.kaks'), header=TRUE)
    a$chrom <- paste0('chr',i)
    a$POP=as.factor(pop)
    kaks[[i]]<- a
  }
  return(do.call(rbind,kaks))
}

sum_diff <- function(pop1GR, pop2GR){
  return( sum( width( intersect( pop1GR, pop2GR) ) ) )
}

calc_diff_mat <- function(popList){
  dm <- matrix(nrow = length(popList), ncol = length(popList))
  rownames(dm) <- names(popList)
  colnames(dm) <- names(popList)
  for(row in 1:length(popList)){
    for(col in 1:length(popList)){
      if(row == col){ dm[row,col] <- 0
      } else {
        dm[row,col] <- 1 / sum_diff(popList[[row]],popList[[col]])
      }
    }
  }
  return(dm)
}


report_all_common_regions <- function(listGR){
  if(length(listGR) == 1){
    return(listGR[[1]])
  }
  intsec <- intersect(listGR[[1]], listGR[[2]])
  if(length(listGR) == 2){
    return(intsec)
  }
  for(i in 3:length(listGR)){
    intsec <- intersect(intsec, listGR[[i]])
  } 
  return(intsec)
}

combine_multi_GR <- function(listGR){
  if(length(listGR) == 1){
    return(listGR[[1]])
  }
  uni <- union(listGR[[1]], listGR[[2]])
  if(length(listGR) == 2){
    return(uni)
  }
  for(i in 3:length(listGR)){
    uni <- union(uni, listGR[[i]])
  } 
  return(reduce(uni))
}

report_unique_to_pop <- function(refGR, list_otherGR){
  return(setdiff(refGR,intersect(refGR, combine_multi_GR(list_otherGR), ignore.strand = TRUE ), ignore.strand = TRUE))
}


lsa.gwas.cluster <- function (gwas_data,settings) {
  # Cluster GWAS data to identify regions of association
  # default eps = 0.2 (200kb)
  lambda_cluster <- function (position, eps = 0.5, minPts = 1) {
    dbscan::dbscan(as.matrix(position)/1000000,eps=eps, minPts = minPts)
  }
  
  gwas_data[, ClusterID := lambda_cluster(chrom_start),by= c("chrom","POP")]
  return (gwas_data)
} 

```

```{r load data}
nzcau <-list(faw = get_faw(paste0(data_dir,"faywuh/"),"nzcau"), 
             td = get_td(paste0(data_dir, "tajimasd/"), 'nzcau'), 
             ihs = get_ihs(paste0(data_dir,"ihs/nzcau/"), 'nzcau'), 
             nsl = get_nsl(paste0(data_dir,"nsl/nzcau/"), 'nzcau'),
             kaks = get_kaks(paste0(data_dir,"kaks/"), 'nzcau')
)

nzm <-list(faw = get_faw(paste0(data_dir,"faywuh/"),"nzm"), 
           td = get_td(paste0(data_dir, "tajimasd/"), 'nzm'), 
           ihs = get_ihs(paste0(data_dir,"ihs/nzm/"), 'nzm'), 
           nsl = get_nsl(paste0(data_dir,"nsl/nzm/"), 'nzm'),
           kaks = get_kaks(paste0(data_dir,"kaks/"), 'nzm')
)

cim <-list(faw = get_faw(paste0(data_dir,"faywuh/"),"cim"), 
           td = get_td(paste0(data_dir, "tajimasd/"), 'cim'), 
           ihs = get_ihs(paste0(data_dir,"ihs/cim/"), 'cim'), 
           nsl = get_nsl(paste0(data_dir,"nsl/cim/"), 'cim'),
           kaks = get_kaks(paste0(data_dir,"kaks/"), 'cim'))

ton <-list(faw = get_faw(paste0(data_dir,"faywuh/"),"ton"), 
           td = get_td(paste0(data_dir, "tajimasd/"), 'ton'), 
           ihs = get_ihs(paste0(data_dir,"ihs/ton/"), 'ton'), 
           nsl = get_nsl(paste0(data_dir,"nsl/ton/"), 'ton'),
           kaks = get_kaks(paste0(data_dir,"kaks/"), 'ton'))

sam <-list(faw = get_faw(paste0(data_dir,"faywuh/"),"sam"), 
           td = get_td(paste0(data_dir, "tajimasd/"), 'sam'), 
           ihs = get_ihs(paste0(data_dir,"ihs/sam/"), 'sam'), 
           nsl = get_nsl(paste0(data_dir,"nsl/sam/"), 'sam'),
           kaks = get_kaks(paste0(data_dir,"kaks/"), 'sam'))
```

```{r gwas catalogue, echo = FALSE}
gwas_cat <- read.delim('~/Downloads/gwas_catalog_v1.0.1-associations_e84_r2016-07-10.tsv', header=TRUE, stringsAsFactors = FALSE, sep='\t')
#filter gwas catalogue for results of genome wide significance
gwas_cat <- gwas_cat[ gwas_cat$P.VALUE < 5e-8,]
gwas_interested <- gwas_cat[grep('metabolic syndrome|obesity|diabetes|urate|gout|body mass|lipid traits', gwas_cat$DISEASE.TRAIT, ignore.case = TRUE) ,]
#filter out potentially misleading gwas hits
gwas_interested <- gwas_interested[grep("erectile|lean|autoantibodies|gestational|cancer|psychopharmacol|metaformin|obstructive|interaction|asthmatics|omega|pain|cataracts|time|bilirubin|chain|thyroid|zhi", gwas_interested$DISEASE.TRAIT, invert = TRUE, ignore.case = TRUE),]

gwas_genes <-sort(unique(unlist(strsplit(gwas_interested$REPORTED.GENE.S., split = ', '))))

gwas_genes_entrez <- na.omit(select(org.Hs.eg.db, keys = gwas_genes, columns = c('SYMBOL','ENTREZID'), keytype="SYMBOL"))

gwas_genes_ucsc <- merge(select(TxDb.Hsapiens.UCSC.hg19.knownGene , columns = c("TXID","TXCHROM","TXSTART","TXEND","TXSTRAND"),keys = gwas_genes_entrez$ENTREZID, keytype="GENEID"), gwas_genes_entrez, by.x ='GENEID', by.y = "ENTREZID")
gwas_genes_ucscGR <- GRanges(gwas_genes_ucsc[!is.na(gwas_genes_ucsc$TXID),])

urate_gout <- read.csv('~/Git_repos/UkBioBank/Gout_GWAS/kottgen_table1.csv', stringsAsFactors = FALSE, header= TRUE)
urate_gout <- unlist(strsplit(unique(urate_gout$closest_gene,urate_gout$grail_gene),' '))
urate_goutGR <- gwas_genes_ucscGR[gwas_genes_ucscGR$SYMBOL %in% urate_gout,]

obesity_GR <- gwas_genes_ucscGR[gwas_genes_ucscGR$SYMBOL %in% unique(unlist(strsplit(gwas_interested[grep("obesity|body mass",gwas_interested$DISEASE.TRAIT, ignore.case = TRUE),]$REPORTED.GENE.S., ', '))),]

diabetes_GR <- gwas_genes_ucscGR[gwas_genes_ucscGR$SYMBOL %in% unique(unlist(strsplit(gwas_interested[grep("diabetes",gwas_interested$DISEASE.TRAIT, ignore.case = TRUE),]$REPORTED.GENE.S., ', '))),]
```

```{r, eval = TRUE, echo=FALSE}
annotate_genes <- function(GR){
  target <- GRanges(seqnames(GR), IRanges(GR@ranges))
  loc <- locateVariants(target, TxDb.Hsapiens.UCSC.hg19.knownGene, AllVariants(), ignore.strand)
  loc <- loc[!is.na(loc$GENEID),]
  if(nrow(as.data.frame(loc, row.names = NULL)) > 0){
    colnames(org.Hs.eg.db)
    cols <- c("SYMBOL")
    keys <- na.omit(unique(loc$GENEID))
    symbols <- select(org.Hs.eg.db, keys, cols, keytype="ENTREZID")
    #convert to gene symbol from entrez id
    n <- merge(as.data.frame(loc, row.names = NULL), symbols, by.x = "GENEID", by.y = "ENTREZID", all.x=TRUE)
    return(n)
  } else {
    return(as.data.frame(loc, row.names = NULL))
  }
}

```





## combined boxplot
```{r}
combined_faw <- rbind(nzcau[['faw']], nzm[['faw']], cim[['faw']], ton[['faw']], sam[['faw']])
ggplot(combined_faw, aes(y = FayWu_H, x = factor(chrom))) + geom_boxplot(aes(colour = factor(chrom))) + facet_wrap(~POP)

```

```{r faywuh clustering}
nzcau_faw <- nzcau[['faw']]  %>% arrange(FayWu_H) 
nzcau_faw <- nzcau_faw[1:(0.01 * nrow(nzcau_faw)), ]
nzcau_fawGR <- GRanges(seqnames = nzcau_faw$chrom, IRanges(start = nzcau_faw$chrom_start, end = nzcau_faw$chrom_end), fawh = nzcau_faw$FayWu_H )
print(table(annotate_genes(nzcau_fawGR)$LOCATION))

nzm_faw <- nzm[['faw']]  %>% arrange(FayWu_H) 
nzm_faw <- nzm_faw[1:(0.01 * nrow(nzm_faw)), ]
nzm_fawGR <- GRanges(seqnames = nzm_faw$chrom, IRanges(start = nzm_faw$chrom_start, end = nzm_faw$chrom_end), fawh = nzm_faw$FayWu_H )
print(table(annotate_genes(nzm_fawGR)$LOCATION))

cim_faw <- cim[['faw']]  %>% arrange(FayWu_H) 
cim_faw <- cim_faw[1:(0.01 * nrow(cim_faw)), ]
cim_fawGR <- GRanges(seqnames = cim_faw$chrom, IRanges(start = cim_faw$chrom_start, end = cim_faw$chrom_end), fawh = cim_faw$FayWu_H )
print(table(annotate_genes(cim_fawGR)$LOCATION))

ton_faw <- ton[['faw']]  %>% arrange(FayWu_H) 
ton_faw <- ton_faw[1:(0.01 * nrow(ton_faw)), ]
ton_fawGR <- GRanges(seqnames = ton_faw$chrom, IRanges(start = ton_faw$chrom_start, end = ton_faw$chrom_end), fawh = ton_faw$FayWu_H )
print(table(annotate_genes(ton_fawGR)$LOCATION))

sam_faw <- sam[['faw']]  %>% arrange(FayWu_H) 
sam_faw <- sam_faw[1:(0.01 * nrow(sam_faw)), ]
sam_fawGR <- GRanges(seqnames = sam_faw$chrom, IRanges(start = sam_faw$chrom_start, end = sam_faw$chrom_end), fawh = sam_faw$FayWu_H )
print(table(annotate_genes(sam_fawGR)$LOCATION))

pops <- c('nzcau','nzm','cim','ton','sam')
popList <- list(nzcau = nzcau_fawGR,
                nzm = nzm_fawGR, 
                cim = cim_fawGR,
                ton = ton_fawGR,
                sam = sam_fawGR
)

diff_mat <-calc_diff_mat(popList)
plot(hclust(dist(diff_mat, method="euclidean"), method = "single"))
heatmap(diff_mat, distfun = function(y) dist(y, method="euclidean"))
```


# Tajima's D


## combined
## combined boxplot
```{r}
combined_td <- rbind(nzcau[['td']], nzm[['td']], cim[['td']], ton[['td']], sam[['td']])
ggplot(combined_td, aes(y = TajimaD, x = factor(chrom))) + geom_boxplot(aes(colour = factor(chrom))) + facet_wrap(~POP)

```


```{r td clustering}
nzcau_td <- nzcau[['td']]  %>% arrange(TajimaD) 
nzcau_td <- nzcau_td[1:(0.01 * nrow(nzcau_td)), ]
nzcau_tdGR <- GRanges(seqnames = nzcau_td$chrom, IRanges(start = nzcau_td$chrom_start, end = nzcau_td$chrom_end), td = nzcau_td$TajimaD )
print(table(annotate_genes(nzcau_tdGR)$LOCATION))

nzm_td <- nzm[['td']]  %>% arrange(TajimaD) 
nzm_td <- nzm_td[1:(0.01 * nrow(nzm_td)), ]
nzm_tdGR <- GRanges(seqnames = nzm_td$chrom, IRanges(start = nzm_td$chrom_start, end = nzm_td$chrom_end), td = nzm_td$TajimaD )
print(table(annotate_genes(nzm_tdGR)$LOCATION))

cim_td <- cim[['td']]  %>% arrange(TajimaD) 
cim_td <- cim_td[1:(0.01 * nrow(cim_td)), ]
cim_tdGR <- GRanges(seqnames = cim_td$chrom, IRanges(start = cim_td$chrom_start, end = cim_td$chrom_end), td = cim_td$TajimaD )
print(table(annotate_genes(cim_tdGR)$LOCATION))

ton_td <- ton[['td']]  %>% arrange(TajimaD) 
ton_td <- ton_td[1:(0.01 * nrow(ton_td)), ]
ton_tdGR <- GRanges(seqnames = ton_td$chrom, IRanges(start = ton_td$chrom_start, end = ton_td$chrom_end), td = ton_td$TajimaD )
print(table(annotate_genes(ton_tdGR)$LOCATION))

sam_td <- sam[['td']]  %>% arrange(TajimaD) 
sam_td <- sam_td[1:(0.01 * nrow(sam_td)), ]
sam_tdGR <- GRanges(seqnames = sam_td$chrom, IRanges(start = sam_td$chrom_start, end = sam_td$chrom_end), td = sam_td$TajimaD )
print(table(annotate_genes(sam_tdGR)$LOCATION))

pops <- c('nzcau','nzm','cim','ton','sam')
popList <- list(nzcau = nzcau_tdGR,
                nzm = nzm_tdGR, 
                cim = cim_tdGR,
                ton = ton_tdGR,
                sam = sam_tdGR
)

diff_mat <-calc_diff_mat(popList)
plot(hclust(dist(diff_mat, method="euclidean"), method = "single"))
heatmap(diff_mat, distfun = function(y) dist(y, method="euclidean"))
```

# iHS


##combined
```{r}
combined_ihs <- rbind(nzcau[['ihs']],nzm[['ihs']],cim[['ihs']], ton[['ihs']], sam[['ihs']])  %>%  filter(significant ==1)

ggplot(combined_ihs, aes(x = chrom_start, y = abs(std_ihs), color = as.factor(POP), alpha=0.3)) + geom_point() + facet_wrap(~chrom, scales = 'free_x')
png(filename = 'ihs_combined.png', width = 2000, height = 1500)
last_plot()
dev.off()

```

```{r}

nzcau_sig_ihsGR <- GRanges(seqnames = nzcau[['ihs']]$chrom, IRanges(start = nzcau[['ihs']]$chrom_start, end = nzcau[['ihs']]$chrom_end), significant_ihs = nzcau[['ihs']]$significant)
print(table(annotate_genes(nzcau_sig_ihsGR)$LOCATION))

nzm_sig_ihsGR <- GRanges(seqnames = nzm[['ihs']]$chrom, IRanges(start = nzm[['ihs']]$chrom_start, end = nzm[['ihs']]$chrom_end), significant_ihs = nzm[['ihs']]$significant)
print(table(annotate_genes(nzm_sig_ihsGR)$LOCATION))

cim_sig_ihsGR <- GRanges(seqnames = cim[['ihs']]$chrom, IRanges(start = cim[['ihs']]$chrom_start, end = cim[['ihs']]$chrom_end), significant_ihs = cim[['ihs']]$significant)
print(table(annotate_genes(cim_sig_ihsGR)$LOCATION))

ton_sig_ihsGR <- GRanges(seqnames = ton[['ihs']]$chrom, IRanges(start = ton[['ihs']]$chrom_start, end = ton[['ihs']]$chrom_end), significant_ihs = ton[['ihs']]$significant)
print(table(annotate_genes(ton_sig_ihsGR)$LOCATION))

sam_sig_ihsGR <- GRanges(seqnames = sam[['ihs']]$chrom, IRanges(start = sam[['ihs']]$chrom_start, end = sam[['ihs']]$chrom_end), significant_ihs = sam[['ihs']]$significant)
print(table(annotate_genes(sam_sig_ihsGR)$LOCATION))

pops <- c('nzcau','nzm','cim','ton','sam')
popList <- list(nzcau = nzcau_sig_ihsGR[nzcau_sig_ihsGR$significant_ihs == 1],
                nzm = nzm_sig_ihsGR[nzm_sig_ihsGR$significant_ihs ==1], 
                cim = cim_sig_ihsGR[cim_sig_ihsGR$significant_ihs ==1],
                ton = ton_sig_ihsGR[ton_sig_ihsGR$significant_ihs ==1],
                sam = sam_sig_ihsGR[sam_sig_ihsGR$significant_ihs ==1]
)

diff_mat <-calc_diff_mat(popList)
plot(hclust(dist(diff_mat, method="euclidean"), method = "single"))
heatmap(diff_mat, distfun = function(y) dist(y, method="euclidean"))



```


```{r ihs clustering, eval = TRUE, echo= FALSE}
library(GenomicRanges)
cau_ihs <- combined_ihs %>% filter(POP == "nzcau" & abs(std_ihs) > 2.5)
nzm_ihs <- combined_ihs %>% filter(POP == "nzm" & abs(std_ihs) > 2.5)
cim_ihs <-combined_ihs %>% filter(POP == "cim" & abs(std_ihs) > 2.5)
ton_ihs <- combined_ihs %>% filter(POP == "ton" & abs(std_ihs) > 2.5)
sam_ihs <- combined_ihs %>% filter(POP == "sam" & abs(std_ihs) > 2.5)



cau_clust <- lsa.gwas.cluster(as.data.table(cau_ihs))
cau_regions <- cau_clust[, .(start = min(chrom_start), end = max(chrom_start)), by = c("chrom","ClusterID")]
cau_regions$POP <- "nzcau"
cau_regions <- GRanges(cau_regions)
cau_ihs_regions <- GRanges(combined_ihs %>% filter(POP == "nzcau" & significant == 1))

nzm_clust <- lsa.gwas.cluster(as.data.table(nzm_ihs))
nzm_regions <- nzm_clust[, .(start = min(chrom_start), end = max(chrom_start)), by = c("chrom","ClusterID")]
nzm_regions$POP <- "nzm"
nzm_regions <- GRanges(nzm_regions)
nzm_ihs_regions <- GRanges(combined_ihs %>% filter(POP == "nzm" & significant == 1))

cim_clust <- lsa.gwas.cluster(as.data.table(cim_ihs ))
cim_regions <- cim_clust[, .(start = min(chrom_start), end = max(chrom_start)), by = c("chrom","ClusterID")]
cim_regions$POP <- "cim"
cim_regions <- GRanges(cim_regions)
cim_ihs_regions <- GRanges(combined_ihs %>% filter(POP == "cim" & significant == 1))

ton_clust <- lsa.gwas.cluster(as.data.table(ton_ihs))
ton_regions <- ton_clust[, .(start = min(chrom_start), end = max(chrom_start)), by = c("chrom","ClusterID")]
ton_regions$POP <- "ton"
ton_regions <- GRanges(ton_regions)
ton_ihs_regions <- GRanges(combined_ihs %>% filter(POP == "ton" & significant == 1))

sam_clust <- lsa.gwas.cluster(as.data.table(sam_ihs))
sam_regions <- sam_clust[, .(start = min(chrom_start), end = max(chrom_start)), by = c("chrom","ClusterID")]
sam_regions$POP <- "sam"
sam_regions <- GRanges(sam_regions)
sam_ihs_regions <- GRanges(combined_ihs %>% filter(POP == "sam" & significant == 1))


```





# nSL


##combined
```{r}
combined_nsl <- rbind(nzcau[['nsl']],nzm[['nsl']],cim[['nsl']], ton[['nsl']], sam[['nsl']])  %>%  filter(significant ==1)

ggplot(combined_nsl, aes(x = chrom_start, y = abs(std_nsl), color = as.factor(POP), alpha=0.3)) + geom_point() + facet_wrap(~chrom, scales = 'free_x')
png(filename = 'nsl_combined.png', width = 2000, height = 1500)
last_plot()
dev.off()

```


```{r nsl clustering, eval = TRUE, echo= FALSE}
library(GenomicRanges)
cau_nsl <- combined_nsl %>% filter(POP == "nzcau" & abs(std_nsl) > 2)
nzm_nsl <- combined_nsl %>% filter(POP == "nzm" & abs(std_nsl) > 2)
cim_nsl <-combined_nsl %>% filter(POP == "cim" & abs(std_nsl) > 2)
ton_nsl <- combined_nsl %>% filter(POP == "ton" & abs(std_nsl) > 2)
sam_nsl <- combined_nsl %>% filter(POP == "sam" & abs(std_nsl) > 2)



cau_clust <- lsa.gwas.cluster(as.data.table(cau_nsl))
cau_regions <- cau_clust[, .(start = min(chrom_start), end = max(chrom_start)), by = c("chrom","ClusterID")]
cau_regions$POP <- "nzcau"
cau_regions <- GRanges(cau_regions)
cau_nsl_regions <- GRanges(combined_nsl %>% filter(POP == "nzcau" & significant == 1))
print(table(annotate_genes(cau_nsl_regions)$LOCATION))

nzm_clust <- lsa.gwas.cluster(as.data.table(nzm_nsl))
nzm_regions <- nzm_clust[, .(start = min(chrom_start), end = max(chrom_start)), by = c("chrom","ClusterID")]
nzm_regions$POP <- "nzm"
nzm_regions <- GRanges(nzm_regions)
nzm_nsl_regions <- GRanges(combined_nsl %>% filter(POP == "nzm" & significant == 1))
print(table(annotate_genes(nzm_nsl_regions)$LOCATION))

cim_clust <- lsa.gwas.cluster(as.data.table(cim_nsl ))
cim_regions <- cim_clust[, .(start = min(chrom_start), end = max(chrom_start)), by = c("chrom","ClusterID")]
cim_regions$POP <- "cim"
cim_regions <- GRanges(cim_regions)
cim_nsl_regions <- GRanges(combined_nsl %>% filter(POP == "cim" & significant == 1))
print(table(annotate_genes(cim_nsl_regions)$LOCATION))

ton_clust <- lsa.gwas.cluster(as.data.table(ton_nsl))
ton_regions <- ton_clust[, .(start = min(chrom_start), end = max(chrom_start)), by = c("chrom","ClusterID")]
ton_regions$POP <- "ton"
ton_regions <- GRanges(ton_regions)
ton_nsl_regions <- GRanges(combined_nsl %>% filter(POP == "ton" & significant == 1))
print(table(annotate_genes(ton_nsl_regions)$LOCATION))

sam_clust <- lsa.gwas.cluster(as.data.table(sam_nsl))
sam_regions <- sam_clust[, .(start = min(chrom_start), end = max(chrom_start)), by = c("chrom","ClusterID")]
sam_regions$POP <- "sam"
sam_regions <- GRanges(sam_regions)
sam_nsl_regions <- GRanges(combined_nsl %>% filter(POP == "sam" & significant == 1))
print(table(annotate_genes(sam_nsl_regions)$LOCATION))


```


```{r}
nzcau_sig_nslGR <- GRanges(seqnames = nzcau[['nsl']]$chrom, IRanges(start = nzcau[['nsl']]$chrom_start, end = nzcau[['nsl']]$chrom_end), significant_nsl = nzcau[['nsl']]$significant)

nzm_sig_nslGR <- GRanges(seqnames = nzm[['nsl']]$chrom, IRanges(start = nzm[['nsl']]$chrom_start, end = nzm[['nsl']]$chrom_end), significant_nsl = nzm[['nsl']]$significant)

cim_sig_nslGR <- GRanges(seqnames = cim[['nsl']]$chrom, IRanges(start = cim[['nsl']]$chrom_start, end = cim[['nsl']]$chrom_end), significant_nsl = cim[['nsl']]$significant)

ton_sig_nslGR <- GRanges(seqnames = ton[['nsl']]$chrom, IRanges(start = ton[['nsl']]$chrom_start, end = ton[['nsl']]$chrom_end), significant_nsl = ton[['nsl']]$significant)

sam_sig_nslGR <- GRanges(seqnames = sam[['nsl']]$chrom, IRanges(start = sam[['nsl']]$chrom_start, end = sam[['nsl']]$chrom_end), significant_nsl = sam[['nsl']]$significant)


pops <- c('nzcau','nzm','cim','ton','sam')
popList <- list(nzcau = nzcau_sig_nslGR[nzcau_sig_nslGR$significant_nsl == 1],
                nzm = nzm_sig_nslGR[nzm_sig_nslGR$significant_nsl ==1], 
                cim = cim_sig_nslGR[cim_sig_nslGR$significant_nsl ==1],
                ton = ton_sig_nslGR[ton_sig_nslGR$significant_nsl ==1],
                sam = sam_sig_nslGR[sam_sig_nslGR$significant_nsl ==1]
)

diff_mat <-calc_diff_mat(popList)
plot(hclust(dist(diff_mat, method="euclidean"), method = "single"))
heatmap(diff_mat, distfun = function(y) dist(y, method="euclidean"))

```

```{r pca, eval = FALSE}
p <- prcomp(x = t(abs(as.data.frame(combined_ihs  %>% filter(POP == 'nzcau')  %>% dplyr::select(std_ihs)))))
```

# common regions


```{r combined GR, eval = TRUE}
nzcau_GR <- list(faw= nzcau_fawGR, td = nzcau_tdGR, ihs = nzcau_sig_ihsGR[nzcau_sig_ihsGR$significant_ihs == 1], nsl = nzcau_sig_nslGR[nzcau_sig_nslGR$significant_nsl == 1] )

nzm_GR <- list(faw= nzm_fawGR, td = nzm_tdGR, ihs = nzm_sig_ihsGR[nzm_sig_ihsGR$significant_ihs == 1], nsl = nzm_sig_nslGR[nzm_sig_nslGR$significant_nsl == 1] )

cim_GR <- list(faw= cim_fawGR, td = cim_tdGR, ihs = cim_sig_ihsGR[cim_sig_ihsGR$significant_ihs == 1], nsl = cim_sig_nslGR[cim_sig_nslGR$significant_nsl == 1] )

ton_GR <- list(faw= ton_fawGR, td = ton_tdGR, ihs = ton_sig_ihsGR[ton_sig_ihsGR$significant_ihs == 1], nsl = ton_sig_nslGR[ton_sig_nslGR$significant_nsl == 1] )

sam_GR <- list(faw= sam_fawGR, td = sam_tdGR, ihs = sam_sig_ihsGR[sam_sig_ihsGR$significant_ihs == 1], nsl = sam_sig_nslGR[sam_sig_nslGR$significant_nsl == 1] )

```

## unique between populations by statistics
### nzcau
```{r unique nzcau,warning=FALSE, message=FALSE}
unique_nzcau_regions = list()
unique_nzcau_regions[['faw']] <- report_unique_to_pop(nzcau_GR[['faw']], list_otherGR = list(nzm_GR[['faw']], cim_GR[['faw']], ton_GR[['faw']],sam_GR[['faw']]))

unique_nzcau_regions[['td']] <- report_unique_to_pop(nzcau_GR[['td']], list_otherGR = list(nzm_GR[['td']], cim_GR[['td']], ton_GR[['td']],sam_GR[['td']]))

unique_nzcau_regions[['ihs']] <- report_unique_to_pop(nzcau_GR[['ihs']], list_otherGR = list(nzm_GR[['ihs']], cim_GR[['ihs']], ton_GR[['ihs']],sam_GR[['ihs']]))

unique_nzcau_regions[['nsl']] <- report_unique_to_pop(nzcau_GR[['nsl']], list_otherGR = list(nzm_GR[['nsl']], cim_GR[['nsl']], ton_GR[['nsl']],sam_GR[['nsl']]))

unique_nzcau_regions_genelists <- list()
for(stat in names(unique_nzcau_regions)){
  print(stat)
  unique_nzcau_regions_genelists[[stat]] <- annotate_genes(unique_nzcau_regions[[stat]])
  print(table(unique_nzcau_regions_genelists[[stat]]$LOCATION))
}
```


### nzm
```{r unique nzm,warning=FALSE, message=FALSE}
unique_nzm_regions = list()
unique_nzm_regions[['faw']] <- report_unique_to_pop(nzm_GR[['faw']], list_otherGR = list(nzcau_GR[['faw']], cim_GR[['faw']], ton_GR[['faw']],sam_GR[['faw']]))

unique_nzm_regions[['td']] <- report_unique_to_pop(nzm_GR[['td']], list_otherGR = list(nzcau_GR[['td']], cim_GR[['td']], ton_GR[['td']],sam_GR[['td']]))

unique_nzm_regions[['ihs']] <- report_unique_to_pop(nzm_GR[['ihs']], list_otherGR = list(nzcau_GR[['ihs']], cim_GR[['ihs']], ton_GR[['ihs']],sam_GR[['ihs']]))

unique_nzm_regions[['nsl']] <- report_unique_to_pop(nzm_GR[['nsl']], list_otherGR = list(nzcau_GR[['nsl']], cim_GR[['nsl']], ton_GR[['nsl']],sam_GR[['nsl']]))

unique_nzm_regions_genelists <- list()
for(stat in names(unique_nzm_regions)){
  unique_nzm_regions_genelists[[stat]] <- annotate_genes(unique_nzm_regions[[stat]])
  print(table(unique_nzm_regions_genelists[[stat]]$LOCATION))
}
```



### cim
```{r unique cim,warning=FALSE, message=FALSE}
unique_cim_regions = list()
unique_cim_regions[['faw']] <- report_unique_to_pop(cim_GR[['faw']], list_otherGR = list(nzm_GR[['faw']], nzcau_GR[['faw']], ton_GR[['faw']],sam_GR[['faw']]))

unique_cim_regions[['td']] <- report_unique_to_pop(cim_GR[['td']], list_otherGR = list(nzm_GR[['td']], nzcau_GR[['td']], ton_GR[['td']],sam_GR[['td']]))

unique_cim_regions[['ihs']] <- report_unique_to_pop(cim_GR[['ihs']], list_otherGR = list(nzm_GR[['ihs']], nzcau_GR[['ihs']], ton_GR[['ihs']],sam_GR[['ihs']]))

unique_cim_regions[['nsl']] <- report_unique_to_pop(cim_GR[['nsl']], list_otherGR = list(nzm_GR[['nsl']], nzcau_GR[['nsl']], ton_GR[['nsl']],sam_GR[['nsl']]))

unique_cim_regions_genelists <- list()
for(stat in names(unique_cim_regions)){
  message(stat)
  unique_cim_regions_genelists[[stat]] <- annotate_genes(unique_cim_regions[[stat]])
  print(table(unique_cim_regions_genelists[[stat]]$LOCATION))
}
```

### ton
```{r unique ton,warning=FALSE, message=FALSE}
unique_ton_regions = list()
unique_ton_regions[['faw']] <- report_unique_to_pop(ton_GR[['faw']], list_otherGR = list(nzm_GR[['faw']], cim_GR[['faw']], nzcau_GR[['faw']],sam_GR[['faw']]))

unique_ton_regions[['td']] <- report_unique_to_pop(ton_GR[['td']], list_otherGR = list(nzm_GR[['td']], cim_GR[['td']], nzcau_GR[['td']],sam_GR[['td']]))

unique_ton_regions[['ihs']] <- report_unique_to_pop(ton_GR[['ihs']], list_otherGR = list(nzm_GR[['ihs']], cim_GR[['ihs']], nzcau_GR[['ihs']],sam_GR[['ihs']]))

unique_ton_regions[['nsl']] <- report_unique_to_pop(ton_GR[['nsl']], list_otherGR = list(nzm_GR[['nsl']], cim_GR[['nsl']], nzcau_GR[['nsl']],sam_GR[['nsl']]))

unique_ton_regions_genelists <- list()
for(stat in names(unique_ton_regions)){
  unique_ton_regions_genelists[[stat]] <- annotate_genes(unique_ton_regions[[stat]])
  print(table(unique_ton_regions_genelists[[stat]]$LOCATION))
}
```

### sam
```{r unique sam, warning=FALSE, message=FALSE}
unique_sam_regions = list()
unique_sam_regions[['faw']] <- report_unique_to_pop(sam_GR[['faw']], list_otherGR = list(nzm_GR[['faw']], cim_GR[['faw']], ton_GR[['faw']],nzcau_GR[['faw']]))

unique_sam_regions[['td']] <- report_unique_to_pop(sam_GR[['td']], list_otherGR = list(nzm_GR[['td']], cim_GR[['td']], ton_GR[['td']],nzcau_GR[['td']]))

unique_sam_regions[['ihs']] <- report_unique_to_pop(sam_GR[['ihs']], list_otherGR = list(nzm_GR[['ihs']], cim_GR[['ihs']], ton_GR[['ihs']],nzcau_GR[['ihs']]))

unique_sam_regions[['nsl']] <- report_unique_to_pop(sam_GR[['nsl']], list_otherGR = list(nzm_GR[['nsl']], cim_GR[['nsl']], ton_GR[['nsl']],nzcau_GR[['nsl']]))

unique_sam_regions_genelists <- list()
for(stat in names(unique_sam_regions)){
  unique_sam_regions_genelists[[stat]] <- annotate_genes(unique_sam_regions[[stat]])
  print(table(unique_sam_regions_genelists[[stat]]$LOCATION))
}
```

## between all statistics by populations

```{r}
nzcau_common <- report_all_common_regions(nzcau_GR)
#nzcau_common_genes <- apply(as.data.frame(nzcau_common),1, function(x){queryUCSC(fromUCSCEnsemblGenes(chromosome = x[1], start = x[2], end = x[3]))})
annotate_genes(nzcau_common)
sum(width(nzcau_common))

nzm_common <- report_all_common_regions(nzm_GR)
annotate_genes(nzm_common)
sum(width(nzm_common))

cim_common <- report_all_common_regions(cim_GR)
annotate_genes(cim_common)
sum(width(cim_common))

ton_common <- report_all_common_regions(ton_GR)
annotate_genes(ton_common)
sum(width(ton_common))

sam_common <- report_all_common_regions(sam_GR)
annotate_genes(sam_common)
sum(width(sam_common))
```



# KaKs


```{r}
table(t(as.data.frame(nzcau[['kaks']]  %>% filter(ka/ks > 2)  %>% dplyr::select(GeneName))) %in% gwas_genes)
table(t(as.data.frame(nzm[['kaks']]  %>% filter(ka/ks > 2)  %>% dplyr::select(GeneName))) %in% gwas_genes)
table(t(as.data.frame(cim[['kaks']]  %>% filter(ka/ks > 2)  %>% dplyr::select(GeneName))) %in% gwas_genes)
table(t(as.data.frame(ton[['kaks']]  %>% filter(ka/ks > 2)  %>% dplyr::select(GeneName))) %in% gwas_genes)
table(t(as.data.frame(sam[['kaks']]  %>% filter(ka/ks > 2)  %>% dplyr::select(GeneName))) %in% gwas_genes)
```

# Gout/Urate
## nzcau
```{r}
as.data.frame(GenomicRanges::subsetByOverlaps(nzcau_fawGR, urate_goutGR),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps( urate_goutGR,nzcau_fawGR),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps(nzcau_tdGR, urate_goutGR),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps(urate_goutGR,nzcau_tdGR ),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps(cau_ihs_regions, urate_goutGR),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps(urate_goutGR,cau_ihs_regions),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps(cau_nsl_regions, urate_goutGR),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps(urate_goutGR,cau_nsl_regions),row.names = NULL)
table(urate_goutGR$SYMBOL %in% t(as.data.frame(nzcau[['kaks']]  %>% filter(ka/ks > 2)  %>% dplyr::select(GeneName))) )
```

## nzm
```{r}
as.data.frame(GenomicRanges::subsetByOverlaps(nzm_fawGR, urate_goutGR),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps( urate_goutGR,nzm_fawGR),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps(nzm_tdGR, urate_goutGR),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps(urate_goutGR,nzm_tdGR ),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps(nzm_ihs_regions, urate_goutGR),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps(urate_goutGR,nzm_ihs_regions),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps(nzm_nsl_regions, urate_goutGR),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps(urate_goutGR,nzm_nsl_regions),row.names = NULL)
table(urate_goutGR$SYMBOL %in% t(as.data.frame(nzm[['kaks']]  %>% filter(ka/ks > 2)  %>% dplyr::select(GeneName))) )
```

## cim
```{r}
as.data.frame(GenomicRanges::subsetByOverlaps(cim_fawGR, urate_goutGR),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps( urate_goutGR,cim_fawGR),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps(cim_tdGR, urate_goutGR),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps(urate_goutGR,cim_tdGR ),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps(cim_ihs_regions, urate_goutGR),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps(urate_goutGR,cim_ihs_regions),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps(cim_nsl_regions, urate_goutGR),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps(urate_goutGR,cim_nsl_regions),row.names = NULL)
table(urate_goutGR$SYMBOL %in% t(as.data.frame(cim[['kaks']]  %>% filter(ka/ks > 2)  %>% dplyr::select(GeneName))) )
```

## ton
```{r}
as.data.frame(GenomicRanges::subsetByOverlaps(ton_fawGR, urate_goutGR),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps( urate_goutGR,ton_fawGR),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps(ton_tdGR, urate_goutGR),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps(urate_goutGR,ton_tdGR ),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps(ton_ihs_regions, urate_goutGR),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps(urate_goutGR,ton_ihs_regions),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps(ton_nsl_regions, urate_goutGR),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps(urate_goutGR,ton_nsl_regions),row.names = NULL)
table(urate_goutGR$SYMBOL %in% t(as.data.frame(ton[['kaks']]  %>% filter(ka/ks > 2)  %>% dplyr::select(GeneName))) )
```

## sam
```{r}
as.data.frame(GenomicRanges::subsetByOverlaps(sam_fawGR, urate_goutGR),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps( urate_goutGR,sam_fawGR),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps(sam_tdGR, urate_goutGR),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps(urate_goutGR,sam_tdGR ),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps(sam_ihs_regions, urate_goutGR),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps(urate_goutGR,sam_ihs_regions),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps(sam_nsl_regions, urate_goutGR),row.names = NULL)
as.data.frame(GenomicRanges::subsetByOverlaps(urate_goutGR,sam_nsl_regions),row.names = NULL)
table(urate_goutGR$SYMBOL %in% t(as.data.frame(sam[['kaks']]  %>% filter(ka/ks > 2)  %>% dplyr::select(GeneName))) )
```

